Rework maintenance loot spawners to fire at roundstart

coiax opened the pull request:
- Maintenance loot spawners now fire at roundstart, rather than during
  the initialization of the Atoms subsystem.
- Maintenance loot spawners do not qdel themselves during the spawning
  process, but persist, and merely hide themselves from living players.

This is prepatory refactoring of maintenance loot spawning, to
accomplish the following proposed ideas. Note that none of these ideas
are actually done in this commit.

- Later events that spawn more maintenance loot using the same spawners.
- Allowing admins to add/remove traits like "Filled up maintenance" in
  pre-round and have them take effect.

As a cute side effect, this lets preround observers look at the
colourful maintenance loot dice icons before the round starts. Observers
can continue to see the loot dice post-round, but they are deliberately
very faint.

Cyberboss commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -590,3 +591,8 @@ GLOBAL_LIST_EMPTY(the_station_areas)
 	return isolated_ruins_z.z_value
+
+/datum/controller/subsystem/mapping/proc/spawn_maintenance_loot()
+	for(var/obj/effect/spawner/lootdrop/maintenance/spawner as anything in GLOB.maintenance_loot_spawners)
```
CHECK_TICK this please

Cyberboss commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -590,3 +591,10 @@ GLOBAL_LIST_EMPTY(the_station_areas)
+		CHECK_TICK
+
+		spawner.spawn_loot()
+		spawner.invisibility = INVISIBILITY_ABSTRACT
```
I feel like this should be a property of `/obj/effect/spawner/lootdrop/maintenance`

coiax replied:
I wanted them to be visible in pre-round, as a sort of "hey look guys at the map you live in" for the observers, and being helpful for debugging when I was working with them.

If you feel strongly I can change it, but I thought it was cute. 🤷

coiax replied:
I have added a `hide()` proc to the maintenance loot spawners, which hides from non-observers and set the transparency to low.

AnturK commented:
How does this work with spawners on templates

coiax commented:
> How does this work with spawners on templates

Good catch, it doesn't.

coiax committed:
Rework maintenance loot spawners to fire at roundstart and persist

- Maintenance loot spawners now fire at roundstart, rather than during
  the initialization of the Atoms subsystem.
- Maintenance loot spawners do not qdel themselves during the spawning
  process, but persist, and merely hide themselves from living players.

This is prepatory refactoring of maintenance loot spawning, to
accomplish the following proposed ideas. Note that none of these ideas
are actually done in this commit.

- Later events that spawn more maintenance loot using the same spawners.
- Allowing admins to add/remove traits like "Filled up maintenance" in
  pre-round and have them take effect.

As a cute side effect, this lets preround observers look at the
colourful maintenance loot dice icons before the round starts. Observers
can continue to see the loot dice post-round, but they are deliberately
very faint.

Rohesie commented on code/game/objects/effects/spawners/lootdrop.dm:
```diff
@@ -240,18 +254,52 @@
-	. = ..()
+	return effective_lootcount
+
+/obj/effect/spawner/lootdrop/maintenance/spawn_loot()
+	. = ..(get_effective_lootcount())
```
```suggestion
/obj/effect/spawner/lootdrop/maintenance/spawn_loot(lootcount_override)
	if(!isnull(lootcount_override))
		lootcount_override = get_effective_lootcount()
	return ..()
```
This makes it more clear what is happening (the proc args get passed down the line) and allows admins to use the verb with a custom argument.

coiax replied:
But this also triggers the closet taking-contents on lines 299-302, which needs to happen after calling parent.

Can I keep the `. = ..(lootcount_override)`

Rohesie replied:
Oh, I missed that part. Let me suggestion edited.

Rohesie commented on code/game/objects/effects/spawners/lootdrop.dm:
```diff
@@ -240,18 +254,52 @@
+
+/obj/effect/spawner/lootdrop/maintenance/Destroy()
+	GLOB.maintenance_loot_spawners -= src
+	. = ..()
```
```suggestion
	return ..()
```

Rohesie commented on code/game/objects/effects/spawners/lootdrop.dm:
```diff
@@ -2,13 +2,28 @@
+
+///If the spawner has any loot defined, randomly picks some and spawns it. Does not cleanup the spawner.
+/obj/effect/spawner/lootdrop/proc/spawn_loot(lootcount_override)
+	var/lootcount = src.lootcount
+
+	if(!isnull(lootcount_override))
+		lootcount = lootcount_override
```
```suggestion
	var/lootcount = isnull(lootcount_override) ? src.lootcount : lootcount_override
```

coiax committed:
Update code/game/objects/effects/spawners/lootdrop.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

coiax committed:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>
